BJS2-17595

Caveri95 opened the pull request:
- Добавлена интеграция с сервисом diceBear для получения рандомной аватарки при создании профиля
- Добавлена сервисная логика на создание/сохранение юзера
- Добавлен контроллер на создание юзера
- Написаны тесты

Caveri95 committed:
BJS2-17595

- Добавлена интеграция с сервисом diceBear для получения рандомной аватарки при создании профиля
- Добавлена сервисная логика на создание/сохранение юзера
- Добавлен контроллер на создание юзера
- Написаны тесты

Caveri95 committed:
BJS2-17595

- Вынес логику по работе с аватарами в отдельный сервис

Caveri95 committed:
BJS2-17595

- Добавил интеграцию с БД s3
- Добавлен функционал по рандомному назначению аватарки

ravshanbekn commented on src/main/java/school/faang/user_service/config/AmazonS3Config.java:
```diff
@@ -0,0 +1,34 @@
+
+        if (!amazonS3.doesBucketExistV2(amazonCredentials.getBucketName())) {
+            amazonS3.createBucket(amazonCredentials.getBucketName());
+        }
```
Хорошо подметил)

Caveri95 replied:
Это с документации пример конфигурации клиента s3

ravshanbekn commented on src/main/java/school/faang/user_service/service/AvatarService.java:
```diff
@@ -0,0 +1,66 @@
+        List<String> styles = styleAvatarConfig.getStyles();
+        Collections.shuffle(styles);
+
+        String newStyle = styles.iterator().next();
```
Не лучше ли просто брать случайный индекс styles?

Caveri95 replied:
Да, хотел как бы поменьше написать кода, а получилось, что всегда одно значение будет. Поправлю. Спасибо за ревью:)

ravshanbekn commented on src/main/java/school/faang/user_service/service/AvatarService.java:
```diff
@@ -0,0 +1,66 @@
+
+        String newStyle = styles.iterator().next();
+        smallAvatarUrl = url + newStyle + smallPostfix;
+        fullAvatarUrl = url + newStyle + fullPostfix;
```
Получается при одном запуске берется один случайный style и он на все приложение, пока его не перезапустить?
Если я не ошибаюсь, тут scope singleton

Caveri95 replied:
Хорошее замечание, получается так, подумаю как лучше сделать для рандома для каждого вызова метода

ravshanbekn commented on src/main/java/school/faang/user_service/config/AmazonS3Config.java:
```diff
@@ -0,0 +1,34 @@
+
+        AmazonS3 amazonS3 = AmazonS3ClientBuilder.standard()
+                .withCredentials(new AWSStaticCredentialsProvider(credentials))
+                .withRegion(Regions.US_EAST_1)
```
В application.yml такая строчка есть

Caveri95 replied:
Согласен, переделаю

ravshanbekn reviewed (commented):
Красиво реализовал

Caveri95 committed:
BJS2-17595

- Регион теперь достается из конфига
- Аватарки по дефолту присваиваются рандомно

pablo023 commented on src/main/resources/application.yaml:
Ой ну какой молодец, а)

pablo023 reviewed (commented):
Покажи мне работу этой задачи на демо. После - пофиксишь конфликты, и вмержишь.

Caveri95 committed:
BJS2-17595

- Исправлено добавление ключа при сохранении аватара в БД
- Изменен формат аватара, полученного с diceBear

Kostahe commented on src/main/java/school/faang/user_service/service/AmazonS3Service.java:
```diff
@@ -0,0 +1,36 @@
+        ObjectMetadata objectMetadata = new ObjectMetadata();
+        objectMetadata.setContentLength(file.length);
+        objectMetadata.setContentType(IMAGE_JPEG.getMimeType());
+        PutObjectRequest putObjectRequest =
+                new PutObjectRequest(
+                        amazonCredentials.getBucketName(),
+                        key,
+                        new ByteArrayInputStream(file),
+                        objectMetadata
+                );
+        amazonS3.putObject(putObjectRequest);
```
Тут можно обойтись без создания объекта `PutObjectRequest`. Пример:
```java
amazonS3.putObject(bucketName, fileName, inputStream, objectMetadata);
```
А так норм

Caveri95 replied:
Да, можно и так

Kostahe commented on src/main/java/school/faang/user_service/service/AvatarService.java:
```diff
@@ -0,0 +1,68 @@
+@RequiredArgsConstructor
+public class AvatarService {
+
+    @Value("${services.diceBear.avatarUrl.url}")
+    @Setter
+    private String url;
+
+    @Value("${services.diceBear.avatarUrl.smallPostfix}")
+    @Setter
+    private String smallPostfix;
+
+    @Value("${services.diceBear.avatarUrl.fullPostfix}")
+    @Setter
+    private String fullPostfix;
+
+    @Value("${services.diceBear.avatarUrl.avatarPattern}")
+    @Setter
+    private String avatarPattern;
+
+    @Value("${services.diceBear.avatarUrl.smallAvatarPattern}")
+    @Setter
+    private String smallAvatarPattern;
```
Мне кажется, это можно вынести в отдельный класс Property.

Kostahe commented on src/main/java/school/faang/user_service/service/AvatarService.java:
```diff
@@ -0,0 +1,68 @@
+@RequiredArgsConstructor
+public class AvatarService {
+
+    @Value("${services.diceBear.avatarUrl.url}")
+    @Setter
+    private String url;
+
+    @Value("${services.diceBear.avatarUrl.smallPostfix}")
+    @Setter
+    private String smallPostfix;
+
+    @Value("${services.diceBear.avatarUrl.fullPostfix}")
+    @Setter
+    private String fullPostfix;
+
+    @Value("${services.diceBear.avatarUrl.avatarPattern}")
+    @Setter
+    private String avatarPattern;
+
+    @Value("${services.diceBear.avatarUrl.smallAvatarPattern}")
+    @Setter
+    private String smallAvatarPattern;
```
Не понимаю, зачем здесь используется @Setter. Но можно просто над классом аннотацию `@Setter`

Caveri95 replied:
Да, это не вопрос

Kostahe commented on src/main/java/school/faang/user_service/service/AvatarService.java:
```diff
@@ -0,0 +1,68 @@
+                .build());
+    }
+
+    private String getStyle() {
+        List<String> styles = styleAvatarConfig.getStyles();
+        Collections.shuffle(styles);
+        String newStyle = styles.get(0);
+        return url + newStyle;
```
Почему здесь не используется случайный индекс?
Оптимизировать надо

Caveri95 replied:
да, можно и оптимизировать, спасибо

Kostahe commented on src/main/java/school/faang/user_service/service/RestTemplateService.java:
```diff
@@ -0,0 +1,16 @@
+    private final RestTemplate restTemplate;
+
+    public byte[] getImageBytes(String url) {
+        return restTemplate.getForObject(url, byte[].class);
```
Здесь может возникнуть `RestClientException`

Kostahe reviewed (commented):
Поправь немного, а так круто сделал.

pablo023 commented:
Слушай а давай мы сделаем из титан мастер, твой отдельный мастер, и туда этот пр сделаем? А то в мастере у нас уже есть Равшанбека, но и твой хотелось бы - ты же работал, и хорошую работу сделал. Заодно конфликты все денем куда-то.

Caveri95 committed:
BJS2-17595

- Правки по замечаниям
- Добавлен тест на ошибки при использовании restTemplate

Caveri95 committed:
BJS2-17595

- Запрос на слияние в новую свою ветку от мастера
